# Subject: Classical Data Analysis

## Session 2 - Regression with multiple variables

### Exercise 1 -  Linear Regression with multiple variables 

Considering the Stepwise Regression presented in Demo 1 perform the Linear Regression with multiple variable considering the following independent variables:
    
- Situation 1: radio and newspaper;
- Situation 2: TV, radio and newspaper.
    
Interpret and discuss the OLS Regression Results, namely the different R-squared (R2). 

Commit scripts in your GitHub account. You should export your solution code (.ipynb notebook) and push it to your repository “ClassicalDataAnalysis”.

The following are the tasks that should complete and synchronize with your repository “ClassicalDataAnalysis” until November 10. Please notice that none of these tasks is graded, however it’s important that you correctly understand and complete them in order to be sure that you won’t have problems with further assignments.

# 1. Data

Let's take a look again at some data - Advertising - , ask some questions about that data, and then use linear regression (with one and multiple variables) to answer those questions!

In [5]:
# imports
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
%matplotlib inline

In [6]:
# read data into a DataFrame
df = pd.read_csv('http://www-bcf.usc.edu/~gareth/ISL/Advertising.csv', index_col=0)
df.head()

TV  radio  newspaper  sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

In [7]:
df

TV  radio  newspaper  sales
1    230.1   37.8       69.2   22.1
2     44.5   39.3       45.1   10.4
3     17.2   45.9       69.3    9.3
4    151.5   41.3       58.5   18.5
5    180.8   10.8       58.4   12.9
6      8.7   48.9       75.0    7.2
7     57.5   32.8       23.5   11.8
8    120.2   19.6       11.6   13.2
9      8.6    2.1        1.0    4.8
10   199.8    2.6       21.2   10.6
11    66.1    5.8       24.2    8.6
12   214.7   24.0        4.0   17.4
13    23.8   35.1       65.9    9.2
14    97.5    7.6        7.2    9.7
15   204.1   32.9       46.0   19.0
16   195.4   47.7       52.9   22.4
17    67.8   36.6      114.0   12.5
18   281.4   39.6       55.8   24.4
19    69.2   20.5       18.3   11.3
20   147.3   23.9       19.1   14.6
21   218.4   27.7       53.4   18.0
22   237.4    5.1       23.5   12.5
23    13.2   15.9       49.6    5.6
24   228.3   16.9       26.2   15.5
25    62.3   12.6       18.3    9.7
26   262.9    3.5       19.5   12.0
27   142.9   29.3       12.6   15.0
28   240.1   16.7       22.9   15.9
29   248.8   27.1       22.9   18.9
30    70.6   16.0       40.8   10.5
..     ...    ...        ...    ...
171   50.0   11.6       18.4    8.4
172  164.5   20.9       47.4   14.5
173   19.6   20.1       17.0    7.6
174  168.4    7.1       12.8   11.7
175  222.4    3.4       13.1   11.5
176  276.9   48.9       41.8   27.0
177  248.4   30.2       20.3   20.2
178  170.2    7.8       35.2   11.7
179  276.7    2.3       23.7   11.8
180  165.6   10.0       17.6   12.6
181  156.6    2.6        8.3   10.5
182  218.5    5.4       27.4   12.2
183   56.2    5.7       29.7    8.7
184  287.6   43.0       71.8   26.2
185  253.8   21.3       30.0   17.6
186  205.0   45.1       19.6   22.6
187  139.5    2.1       26.6   10.3
188  191.1   28.7       18.2   17.3
189  286.0   13.9        3.7   15.9
190   18.7   12.1       23.4    6.7
191   39.5   41.1        5.8   10.8
192   75.5   10.8        6.0    9.9
193   17.2    4.1       31.6    5.9
194  166.8   42.0        3.6   19.6
195  149.7   35.6        6.0   17.3
196   38.2    3.7       13.8    7.6
197   94.2    4.9        8.1    9.7
198  177.0    9.3        6.4   12.8
199  283.6   42.0       66.2   25.5
200  232.1    8.6        8.7   13.4

[200 rows x 4 columns]

# 2 - Linear Regression with multiple variables in Statsmodels

## 2.1. Regression model with multiple variables and with a constant for Situation 1

In [9]:
Y=df['sales']
X=df[['radio','newspaper']]

In [10]:
X=sm.add_constant(X)

In [13]:
model = sm.OLS(Y, X).fit()

In [14]:
prediction=model.predict(X)
prediction

1      17.172583
2      17.311025
3      18.785509
4      17.798146
5      11.726622
6      19.420514
7      15.873721
8      13.167267
9       9.613558
10      9.847293
11     10.504168
12     13.992567
13     16.613237
14     10.749497
15     16.043120
16     19.034824
17     17.231388
18     17.441831
19     13.390923
20     14.072990
21     15.057255
22     10.360186
23     12.683281
24     12.726851
25     11.818471
26     10.015138
27     15.104644
28     12.665117
29     14.735181
30     12.644716
         ...    
171    11.620091
172    13.663886
173    13.302668
174    10.687183
175     9.952711
176    19.199928
177    15.334944
178    10.975343
179     9.804190
180    11.296304
181     9.761583
182    10.445812
183    10.520807
184    18.224890
185    13.627896
186    18.296057
187     9.783649
188    15.022424
189    11.980224
190    11.752834
191    17.408189
192    11.378467
193    10.214959
194    17.572712
195    16.314773
196    10.017075
197    10.218057
198    11.0825

In [15]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.333
Model:                            OLS   Adj. R-squared:                  0.326
Method:                 Least Squares   F-statistic:                     49.11
Date:                Fri, 27 Oct 2017   Prob (F-statistic):           4.96e-18
Time:                        12:02:34   Log-Likelihood:                -573.24
No. Observations:                 200   AIC:                             1152.
Df Residuals:                     197   BIC:                             1162.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.1889      0.628     14.640      0.000       7.951      10.427
radio          0.1990      0.022      9.101      0.000       0.156       0.242
newspaper      0.0066      0.015      0.446      0.656      -0.023       0.036
==============================================================================
Omnibus:                       19.261   Durbin-Watson:                   1.943
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               21.768
Skew:                          -0.765   Prob(JB):                     1.88e-05
Kurtosis:                       3.519   Cond. No.                         92.0
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [26]:
model.params

const        9.188920
radio        0.199045
newspaper    0.006644
dtype: float64

## 2.2. Regression model with multiple variables and with a constant for Situation 2

> Put your code here

In [17]:
x1=df[['TV','radio','newspaper']]
y1=df['sales']

In [19]:
x1=sm.add_constant(x1)

In [20]:
model1 = sm.OLS(y1, x1).fit()

In [22]:
prediction1=model1.predict(x1)
prediction1

1      20.523974
2      12.337855
3      12.307671
4      17.597830
5      13.188672
6      12.478348
7      11.729760
8      12.122953
9       3.727341
10     12.550849
11      7.032299
12     17.285129
13     10.577121
14      8.826300
15     18.434366
16     20.819300
17     12.823657
18     23.224957
19      9.951682
20     14.166073
21     18.100767
22     14.740538
23      6.489150
24     16.545933
25      8.146519
26     15.610039
27     14.989514
28     17.051673
29     19.410538
30      9.144024
         ...    
171     7.394980
172    14.358274
173     7.607692
174    11.970939
175    13.744357
176    24.786870
177    19.979373
178    12.162046
179    16.010997
180    12.384555
181    10.587200
182    13.928099
183     6.554670
184    24.133100
185    18.538521
186    20.803011
187     9.691373
188    17.076442
189    18.644306
190     6.051624
191    12.489159
192     8.424019
193     4.466230
194    18.486958
195    16.495300
196     5.370342
197     8.165312
198    12.7859

In [23]:
model1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Fri, 27 Oct 2017   Prob (F-statistic):           1.58e-96
Time:                        12:09:58   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [24]:
model1.params

const        2.938889
TV           0.045765
radio        0.188530
newspaper   -0.001037
dtype: float64

## 2.3. Regression model with multiple variables and without a constant for Situation 1

In [31]:
Y1=df['sales']
X1=df[['radio','newspaper']]

In [32]:
X1.head()


radio  newspaper
1   37.8       69.2
2   39.3       45.1
3   45.9       69.3
4   41.3       58.5
5   10.8       58.4

In [33]:
model2=sm.OLS(Y1, X1).fit()

In [34]:
model2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.831
Model:                            OLS   Adj. R-squared:                  0.830
Method:                 Least Squares   F-statistic:                     487.9
Date:                Fri, 27 Oct 2017   Prob (F-statistic):           3.02e-77
Time:                        12:15:55   Log-Likelihood:                -646.85
No. Observations:                 200   AIC:                             1298.
Df Residuals:                     198   BIC:                             1304.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
radio          0.3763      0.026     14.334      0.000       0.325       0.428
newspaper      0.1024      0.019      5.303      0.000       0.064       0.140
==============================================================================
Omnibus:                       37.865   Durbin-Watson:                   1.653
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               53.382
Skew:                          -1.137   Prob(JB):                     2.56e-12
Kurtosis:                       4.110   Cond. No.                         3.16
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [36]:
model2.params

radio        0.376286
newspaper    0.102399
dtype: float64

In [39]:
prediction2=model2.predict(X1)
prediction2

1      21.309603
2      19.406226
3      24.367761
4      21.530940
5      10.043970
6      26.080292
7      14.748556
8       8.563034
9       0.892600
10      3.149195
11      4.660506
12      9.440464
13     19.955715
14      3.597045
15     17.090153
16     23.365739
17     25.445517
18     20.614777
19      9.587762
20     10.949054
21     15.891214
22      4.325427
23     11.061922
24      9.042081
25      6.615101
26      3.313775
27     12.315409
28      8.628908
29     12.542285
30     10.198443
         ...    
171     6.249055
172    12.718076
173     9.304130
174     3.982334
175     2.620795
176    22.680658
177    13.442536
178     6.539464
179     3.292305
180     5.565078
181     1.828253
182     4.837667
183     5.186070
184    23.532528
185    11.086855
186    18.977522
187     3.514004
188    12.663069
189     5.609253
190     6.949191
191    16.059276
192     4.678283
193     4.778570
194    16.172657
195    14.010181
196     2.805360
197     2.673231
198     4.1548

## 2.4. Regression model with multiple variables and without a constant for Situation 2

In [40]:
X2=df[['TV','radio','newspaper']]
Y2=df['sales']

In [41]:
model3=sm.OLS(Y2,X2).fit()

In [42]:
model3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.982
Model:                            OLS   Adj. R-squared:                  0.982
Method:                 Least Squares   F-statistic:                     3566.
Date:                Fri, 27 Oct 2017   Prob (F-statistic):          2.43e-171
Time:                        12:18:24   Log-Likelihood:                -423.54
No. Observations:                 200   AIC:                             853.1
Df Residuals:                     197   BIC:                             863.0
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0538      0.001     40.507      0.000       0.051       0.056
radio          0.2222      0.009     23.595      0.000       0.204       0.241
newspaper      0.0168      0.007      2.517      0.013       0.004       0.030
==============================================================================
Omnibus:                        5.982   Durbin-Watson:                   2.038
Prob(Omnibus):                  0.050   Jarque-Bera (JB):                7.039
Skew:                          -0.232   Prob(JB):                       0.0296
Kurtosis:                       3.794   Cond. No.                         12.6
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [44]:
model3.params

TV           0.053792
radio        0.222227
newspaper    0.016816
dtype: float64

In [46]:
prediction3=model3.predict(X2)
prediction3

1      21.941397
2      11.885691
3      12.290830
4      18.311216
5      13.107703
6      12.596134
7      10.777274
8      11.016509
9       0.946104
10     11.681915
11      5.251518
12     16.949837
13     10.188625
14      7.054714
15     19.063754
16     22.000762
17     13.697677
18     24.875590
19      8.585798
20     13.555969
21     18.801838
22     14.298736
23      5.077560
24     16.476916
25      6.459038
26     15.247600
27     14.410006
28     17.011721
29     19.790874
30      8.039452
         ...    
171     5.576852
172    14.290411
173     5.806968
174    10.851616
175    12.939181
176    26.464811
177    20.414540
178    11.480687
179    15.793884
180    11.426176
181     9.141175
182    13.414322
183     4.789246
184    26.233735
185    18.890312
186    21.379388
187     8.417960
188    16.963610
189    18.535657
190     4.088362
191    11.355857
192     6.562240
193     2.367750
194    18.366571
195    16.064835
196     3.109157
197     6.292321
198    11.6955

### Interpreting the fitness of the models

The model have more higher R-squared=0.982 and the model calculated use all of variables (TV,radio,newspaper). 
All of variables p value <0.05 and we accept our hypotesis.TV,newspaper and radio advertising effect sales
